pulp_ansible/app/migrations/0044_collectionversion_sha256_migrate.py

```diff
@@ -0,0 +1,62 @@
+# Generated by Django 3.2.14 on 2022-07-21 22:35
+
+from django.db import migrations, transaction
+
+
+def add_sha256_to_current_models(apps, schema_editor):
+    """Adds the sha256 to current CollectionVersion models."""
+    CollectionVersion = apps.get_model('ansible', 'CollectionVersion')
+    ContentArtifact = apps.get_model('core', 'ContentArtifact')
+    collection_bulk = {}
+    collections_to_update = []
+    collections_on_demand = []
+
+    def find_and_update_sha256():
+        # All content have a ContentArtifact
+        content_artifacts = ContentArtifact.objects.filter(content__in=collection_bulk.keys()).select_related("artifact", "content")
+        for content_artifact in content_artifacts:
+            found_collection = collection_bulk[content_artifact.content.pk]
+            # The ContentArtifact could point to an Artifact or be on-demand
+            if (artifact := getattr(content_artifact, "artifact")) and artifact.sha256 is not None:
+                found_collection.sha256 = artifact.sha256
+                collections_to_update.append(found_collection)
+            else:
+                collections_on_demand.append(found_collection)
+        collection_bulk.clear()
+
+    for collection_version in CollectionVersion.objects.only("pk", "sha256").iterator():
+        if not collection_version.sha256:
+            collection_bulk[collection_version.pk] = collection_version
+            if len(collection_bulk) == 1024:
+                find_and_update_sha256()
+            if len(collections_to_update) >= 1024:
+                with transaction.atomic():
+                    CollectionVersion.objects.bulk_update(collections_to_update, ["sha256",])
+                collections_to_update.clear()
+    # Update remaining collections
+    if len(collection_bulk) > 0:
+        find_and_update_sha256()
+        if len(collections_to_update) > 0:
+            with transaction.atomic():
+                CollectionVersion.objects.bulk_update(collections_to_update, ["sha256",])
+
+    # If there are on-demand collections then the next migration will fail, so error here with
+    # helpful message on how to fix. No work will be performed by this migration on a second-run.
+    if len(collections_on_demand) > 0:
+        raise Exception(
```

Comment:
I don't think it is a good practice to raise exception in the middle on the migration because a) it already performed some changes to the db b) rollback might not be easiest in the middle of failed upgrade.

Is it possible to fetch sha256 from the RemoteArtifact of the on_demand collection content?

Reply:
> I don't think it is a good practice to raise exception in the middle on the migration because a) it already performed some changes to the db b) rollback might not be easiest in the middle of failed upgrade.

Well I tried to design the migration to be able to be re-ran without re-doing the work it has already done and reverting this migration should be as easy as simply removing the new sha256 column. 

> Is it possible to fetch sha256 from the RemoteArtifact of the on_demand collection content?

Yes if sha256 was set on the RemoteArtifact, but it isn't required and the GIt sync code does not set the sha256 when doing an on_demand sync.

Reply:
I see.
My suggestion would be to first ( before doing any modifications to the data on the collections) identify whether there are any collections for which it is not possible to retrieve sha256 ( whether from artifact or remote_artifact), then find in which repos they are, and as last, raise error saying what repos need to be synced with immediate policy. Displaying a list of collection.pk is not very friendly and informative.

Reply:
If we don't want anything to change in case we are unable to perform the whole migration, i would simply rely on the atomic transaction a migration is usually run in. Adding the extra check upfront introduces unnecessary complexity.